### DISCUSSION TITLE
Image Format Issue Tracker.

### DISCUSSION THREAD

**Discussion by etemesi254** [etemesi254/zune-image#102] [General]
## Image formats tracker.

This issue tracks currently supported images and gives details on their status and tracking issues.

### Formats currently able to decode
- Decoder
  - JPG
  - PPM
  - PSD - (Experimental)
  - QOI
  - Farbfeld
  

Priority of decoders follows the following criteria.

### First class image formats.

Images  present everywhere and used for everything, and by first class support, we want to have excellent decoding +encoding of almost all images for that format.
Support for such takes priority.

#### 1. JPEG
Full support is about to be complete, what is left is some polishing and some decoder fixes,
works on a lot of images but some cause errors, looking into them.

#### 2. PNG
Not yet working, but work is underway, mainly spending time translating Eric Bigger's libdeflate to Rust, but support should be soon.

#### 3. GIF.
Limited by the fact that I haven't decided how frame API will look, so this will take a while

### Second class image formats.

Image formats on the rise 

#### 4. WEBP + AVIF + JPEG-XL+HEIF

Complicated and complex beasts, support for these formats will take a while, probably a year or so , decoding support should be pure rust(there is no hurry to reach production at all), while encoding should be their C libraries (encoding is another complicated beast and man has 24 hours a day).

# Other file formats.
Support is welcome.

**Comment by etemesi254** [comment]
cc @Shnatsel , PSD decoding landed(limited in scope, not fully working),  thought you might be interested.

**Comment by Shnatsel** [comment]
Is there a way to toggle which decoders are compiled, or some way to toggle them at runtime?

In many cases it would be nice to disable PSD support either to reduce attack surface (e.g. an image format converter accepting untrusted input) or just not to advertise support for a format that is not complete (e.g. in a desktop image viewer). PSD is an immensely complex format, and full support for it is a multi-year effort.

**Comment by Shnatsel** [comment]
PNG also has an animated variant, [APNG](https://en.wikipedia.org/wiki/APNG), which is now [supported by all web browsers](https://caniuse.com/apng) as well as the Rust `png` and `image` crates. It would be nice to support that as well eventually.

**Comment by etemesi254** [comment]
Yea that's on my radar

Webp + avif also have animation. Will be interesting to see how the final
api version looks

**Comment by etemesi254** [comment]
> Is there a way to toggle which decoders are compiled, or some way to toggle them at runtime?
> 
> In many cases it would be nice to disable PSD support either to reduce attack surface (e.g. an image format converter accepting untrusted input) or just not to advertise support for a format that is not complete (e.g. in a desktop image viewer). PSD is an immensely complex format, and full support for it is a multi-year effort.

`zune-image` supports ability to disable , I'm yet to add that to the binary.

For PSD complexity you make an interesing case,

Just found out ffmpeg supports PSD files, i.e https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/psd.c that's their implementation.
ImageMagick's one is longer since it supports more features i.e 

https://github.com/ImageMagick/ImageMagick/blob/31176f90be5bb143302d2a7a2746a601dc9e2ac0/coders/psd.c

The I was using stb as a basis for my PSD implementation. about 246 LOC at https://github.com/nothings/stb/blob/8b5f1f37b5b75829fc72d38e7b5d4bcbf8a26d55/stb_image.h#L6002-L6248.

Currently we throw out a lot of data we don't need, i.e layers and such to just get the image, the same with ffmpeg and stb. Imagemagick does not do this, and the features it implements aren't  

I'd rather not disable it, for the command line and add it under a feature , say `--unsafe` is needed to decode PSD files since it's experimental at best than trying to chase perfection since that will mean this library will never be ready for any real world decoding.

**Comment by Shnatsel** [comment]
Is the BMP format in scope? It should be relatively trivial, but I don't see it mentioned explicitly.

**Comment by etemesi254** [comment]
Hi, BMP support is planned.

The plan is to support at least anything image-rs supports first.(Like farbfeld decode support landed but can't currently be encoded, working on that one over the weekend).

This issue should probably be restructured in the way it is(I.e tracking image status and planned times )

**Comment by etemesi254** [comment]
> Is there a way to toggle which decoders are compiled, or some way to toggle them at runtime?
> 
> In many cases it would be nice to disable PSD support either to reduce attack surface (e.g. an image format converter accepting untrusted input) or just not to advertise support for a format that is not complete (e.g. in a desktop image viewer). PSD is an immensely complex format, and full support for it is a multi-year effort.

Added `is_experimental` method to `DecoderTrait`, (`DecoderTrait` allows unanimous image decoding from multiple formats into a single image, i.e. implemented for all image decoders).
 
And added an `--experimental` flag in the command line, this adds a guard on all decoders which are experimental(currently farbfeld and psd) to require `--experimental` flag in the command line before decoding.

**Comment by etemesi254** [comment]
@Shnatsel , Finished a simple PNG decoder.

Doesn't do a lot yet, only works with  8 bpp pixel images but I think you'll love the speeds at 
https://github.com/etemesi254/zune-image/blob/main/zune-png/Benches.md

Uses a custom zlib decoder at `zune-inflate` which is as fast as `zlib-ng` (for the tests I've done, which is extracting delfate chunks from a png image of your AI upscaled image from before, running decode on it) and about 20Mb/s faster than miniz-oxide(using the same procedure).

The speedups actually don't even come from that.
They come from a carefully written filter algorithm which the compiler can optimize, and that's pretty sweet(will probably write a blog post about it soon) but for now that's all.

Will work on finishing this and the jpeg decoder for now

**Comment by Shnatsel** [comment]
That's excellent news, thanks for sharing!

I'm curious, how does `zune-inflate` compare to `libdeflate` performance-wise?

Also, I should have mentioned it earlier, but [WUFFS has a fast PNG decoder and detailed explanation of why it's fast](https://nigeltao.github.io/blog/2021/fastest-safest-png-decoder.html), and there's also [libspng](https://libspng.org/) that beats libpng. You might be able to lift some tricks from those and push performance even further.

**Comment by etemesi254** [comment]
Perfomance wise with `libdeflate`, `zune-inflate` is probably a 1 to 1 translation of libdeflate with some small language differences, will put up performance comparisions soon, but in face value I expect `libdeflate` to be slightly faster because it can take advantage of some C properties like incrementing pointers,unbound checks,some union hackery  and some wrapping subtractions on steroids, which I didn't implement as it made some asserts hard to do in some places.

Didn't know about `libspng` let me check it out.
Also looking at `wuffs`.

Will cc you when I think it's ready for fuzz-testing

**Comment by Shnatsel** [comment]
FWIW what I've been doing so far is testing on real-world images instead of fuzz testing. Fuzzing produces random/invalid data to trigger unexpected codepaths, like panics or memory safety bugs. 

Fuzzing usually finds a large amount of rather boring panics, so I've put it off until the correctness on real-world images is established.

**Comment by etemesi254** [comment]
Added jpeg encoding support, via [jpeg-encoder](https://docs.rs/jpeg-encoder/latest/jpeg_encoder/), I figured there is no need for me to do it, since it does it quite better than whatever I'll do.

But it allowed me to finally do some testing that makes sense.
I.e converting from png to jpeg.

|decoder|avg time     |
|------------|---------------- |
|zune      | 475.66 ms |
|vips       | 768.77 ms  |
|magick | 926.42  ms |

The command to run was `perf stat -r 10 [COMMAND]` which runs a command 10 times and collects statistics, statistics lifted were `task-clock:u`

**Reply by Shnatsel** [reply]
FYI [`hyperfine`](https://crates.io/crates/hyperfine) is a really handy command-line benchmarking tool.

**Comment by etemesi254** [comment]
@Shnatsel added small support for lossless jpeg-xl support.

Currently uses 4 threads to encode, pretty neat imo

Pinging you to run on corpus, not yet fuzz tested

**Comment by Shnatsel** [comment]
Ah, I see it's the encoder, not decoder. In that case the corpus doesn't matter. I can write a fuzz target, it just won't need any particular corpus. I understand roundtrip fuzzing is not useful yet because there's no JPEG XL decoder?

A multithreaded JPEG XL encoder is certainly interesting, I wish the format had more uptake.

**Reply by etemesi254** [reply]
The format is really complicated, the encoder supports a small subset of it (lossless modular).

The code is also written in heavy templated C++, hence it may or may not translate to direct Rust.

Mainly wrote it to compare png vs jpeg-xl and jpeg-xl wins on most of my tests hence it's pretty neat, and encode speeds are really really good.

The tests I wanted you to run were bit identical tests with the png corpus to ensure the encoder is but identical with the input image

**Comment by etemesi254** [comment]
Made a png  release

https://crates.io/crates/zune-png

**Reply by Shnatsel** [reply]
Nice, I'll write up an announcement in the next few days. Are the benchmarks [here](https://etemesi254.github.io/assets/criterion/report/index.html) up to date? Also, they're comparing against the current stable `png` rather than the experimental fdeflate one, right?

**Reply by etemesi254** [reply]
Yea I also wanted you to include them in the announcements, would be nice if we could sync with them as there was some really good speed improvements 


The benchmarks are outdated, will run soon. Will update benchmarks  soon



Also would you kindly run the png tests again, to ensure the new header readings don't do anything unexpected.


also thinking of doing a small blog post , the vision I had is slowly coming together

**Reply by etemesi254** [reply]
Yea I also wanted you to include them in the announcements, would be nice if we could sync with them as there was some really good speed improvements 


The benchmarks are outdated, will run soon. Will update benchmarks  soon



Also would you kindly run the png tests again, to ensure the new header readings don't do anything unexpected.


also thinking of doing a small blog post , the vision I had is slowly coming together

**Reply by Shnatsel** [reply]
Okay, here's the announcement draft: https://hackmd.io/@xmxSgN6MSImJPcY4MOKfjA/SJ19v-z-2

Just pending the updated benchmark results (from you), and the test results (from me).

**Reply by Shnatsel** [reply]
The tests results are in, everything looks good :+1:

**Reply by etemesi254** [reply]
Just thought of a way to make some routines a bit faster, in a safe way, lemme work on them then make `0.2.0`

 and for where it matters, we beat slightly `wuffs` on plain RGBA png decode on my machine

```text
`Benchmarkrust_png_decode_image_552k_32bpp_verify_checksum            200         551079 ns/op   1000.972 MB/s`

Benchmark wuffs_png_decode_image_552k_32bpp_verify_checksum/gcc12             400          597292 ns/op       923.527 MB/s
```
(might be that it does BGRA and I do RGBA, still don't know why they choose BGRA, but that's nice)

**Reply by Shnatsel** [reply]
If you beat WUFFS you have a credible claim to the world's fastest PNG decoder :sweat_smile:

**Reply by Shnatsel** [reply]
I'd love to put "world's fastest PNG decoder" in the headline, but I'm not comfortable doing that unless benchmarks against WUFFS are also present on the website with all the others.

**Reply by etemesi254** [reply]
It's difficult  to measure performance differences with it, the benchmarks in the site are weird, they convert RGB to BGRA? Even image-rs/png does not have the ability to convert RGB to BGRA.

It seems the only type of images we both agree on how to decode is RGBA, let me file an issue asking for help.

**Reply by etemesi254** [reply]
Updated  benchmarks, we got faster in some cases, spng got a lot faster since I switched to zlib-ng as backend,  png got faster too.

Would you do a last run in corpus to ensure the code is correct?
If it is, I can release 0.2.0

**Reply by Shnatsel** [reply]
Which `png` crate version are you comparing against now?

**Comment by etemesi254** [comment]
the 0.17-8-rc with new optimizations

**Reply by Shnatsel** [reply]
`zune-png` looks good on my corpus and an additional 58k images I scraped off the web for testing. Excellent!

**Comment by etemesi254** [comment]
@Shnatsel  made 0.2 https://crates.io/crates/zune-png, 

Feel free to announce whenever you are ready

**Reply by Shnatsel** [reply]
I'll announce on Monday, that seems to be the best time

**Reply by Shnatsel** [reply]
PNG announcement posted: https://redd.it/12hlshh

**Comment by jpochyla** [comment]
Hi everybody, thanks a lot for all the impressive work you've done here. I was wondering if PNG encoding is planned?

**Reply by etemesi254** [reply]
hi, there is already a png encoder, it works, just not good enough yet. The only problem is that the compressor(inflate) doesn't compress anything yet so it means it leaves most of the data uncompressed, and generates really huge images.

But it is there and working

**Comment by Quackdoc** [comment]
I was wondering if zune-image would be on a "off by default basis" due to it's complexity (and not yet quite ready state) be willing to implement avif decoding using the rav1d crate in the future?

I understand that it goes against the "minimal" and even possibly "no large crates" (also the decoder is not yet "safe"). fully understand if it's a hard no as it is a rather large crate, and you also need to wind up using an mp4 parsing crate. but it could be nice to have, I ask because it seems like jxl-oxide was implemented for decoding jxl images. 

https://github.com/memorysafety/rav1d
https://github.com/mozilla/mp4parse-rust

**Reply by etemesi254** [reply]
Nb: You can do it yourself for specific crates. E.g see how rimage does for avif https://github.com/SalOne22/rimage/blob/main/src/codecs/avif/encoder/mod.rs.
